[stable/concourse] Refactor values and make secrets optional

william-tran opened the pull request:
Addresses #3208 with the original intent of the issue, to eliminate config map as a source of duplication and make secrets in helm values optional. I could take another stab at it using configmap and secrets supplying an envFrom, but I find this approach simpler.

This PR includes the commits in #3203 assuming that'll be merged first. Only the latest commit in this PR needs to be considered.

Commit:
Make secrets optional, eliminate configmap.

Provide a way to opt out of managing secrets in helm values.

The configmap introduced needless repetition, mapping values directly to
environment variables is a simpler approach.

william-tran commented:
/assign @viglesiasce @nexeck @mattfarina

viglesiasce commented:
/assign

mattfarina commented on stable/concourse/README.md:
```diff
@@ -162,58 +162,43 @@ $ helm install --name my-release -f values.yaml stable/concourse
+### Secrets
 
-To run Concourse securely you'll need [3 private keys](https://concourse.ci/binaries.html#generating-keys). For your convenience, this chart provides some [default keys](concourse-keys), but it is recommended that you generate your own keys by running:
+For your convenience, this chart provides some default values for secrets, but it is recommended that you generate and manage these secrets outside the Helm chart. To do this, set `secrets.create` to `false`, create files for each secret value, and turn it all into a k8s secret. First, perform the following to create the manditory secret values:
```
@viglesiasce any idea how we run automated testing on this if mandatory secret values need to be manually created?

It would be nice if you could pass in a secret to use that was created outside the chart. If that didn't happen the chart could generate a secret.

william-tran replied:
Right now automated testing would only cover the case where `secrets.create: true,` as in the values defaults, see https://github.com/kubernetes/charts/pull/3254/files#diff-95cc28250f44d26990fee96ad2fbd63dR465 and below (diff not rendered automatically by github)

mattfarina commented on stable/concourse/concourse-keys/host_key:
```diff
@@ -1,27 +0,0 @@
-FS4HAoGAGy+56+zvdROjJy9A2Mn/4BvWrsu4RSQILBJ6Hb4TpF46p2fn0rwqyhOj
-Occs+xkdEsI9w5phXzIEeOq2LqvWHDPxtdLpxOrrmx4AftAWdM8S1+OqTpzHihK1
-y1ZOrWfvON+XjWFFAEej/CpQZkNUkTzjTtSC0dnfAveZlasQHdI=
------END RSA PRIVATE KEY-----
```
FYI, templates can generate keys like this. See https://masterminds.github.io/sprig/crypto.html

william-tran replied:
Cool, though I don't see a function that can generate both the public key and private key, or extract a public key given a private key.

viglesiasce commented:
/ok-to-test

viglesiasce commented:
Not sure why but in the test run and when running this locally the pods stayed in `ImagePullBackOff` for minutes. Going to retest in a bit as maybe its just a temporary Docker Hub issue.

william-tran commented:
/test all

william-tran commented:
I think that was a blip, reran tests and it was ready in 30 seconds

ilyasotkov commented:
You should definitely add a notice to README that manually created secrets must NOT have a trailing newline (`\n`). 

Otherwise, running `kubectl create secret generic my-release-concourse --from-file=.` will lead to secrets and eventually container environmental variables like `CONCOURSE_GITHUB_AUTH_CLIENT_ID` having a newline at the end when there really shouldn't be one.

Just spend a few hours debugging this and wouldn't wish it upon anyone :)

william-tran commented:
I'm so sorry @ilyasotkov, I hope f748073 addresses that.

ilyasotkov commented:
Also noticed, this block in templates/secrets.yaml should have `postgres-password`, `postgresql-user` keys too:
```yaml
  {{ if .Values.postgresql.enabled }}
  postgresql-user: {{ .Values.postgresql.postgresUser | b64enc | quote }}
  {{ else }}
  postgresql-uri: {{ .Values.secrets.postgresql.uri | b64enc | quote }}
  {{ end }}
```

viglesiasce commented:
@william-tran do you have a sec to look at the isse that @ilyasotkov found wrt `postgres-password` and `postgres-user`

william-tran commented:
@ilyasotkov I don't think the postgres-password should be in these secrets as it needs to end up in the secrets for the postgres chart dependency as well, see https://github.com/autonomic-ai/charts/blob/ce0404a6a87c738baad46794c3ea630a93288429/stable/concourse/templates/web-deployment.yaml#L49

If you are using the postgres chart dependency, there's no way to keep that password out of your helm values. But now I question whether the postgres username needs to be a secret, or whether that can just go into the values. This was the way it was set up before I added the conditional postgres dependency and kept this structure, with the username in these secrets and the password in the postgres chart's secrets. But the postgres chart doesn't treat the username as a secret: https://github.com/kubernetes/charts/blob/master/stable/postgresql/templates/deployment.yaml#L34

So I'll further simplify this.

Commit:
Don't store postgresql user as a secret

william-tran commented:
@viglesiasce I've addressed @ilyasotkov's concern in the latest commit e8a9a96

ilyasotkov commented:
@william-tran you are right about `postgres-password` being handled by the postgres subchart, there seems to be no way to keep it out of release values as of now, though the subchart will generate a password for us by default if we don’t put one in our release values.

Also I think some massive confusion happened on my end there as I recommended you add the `postresql-user` key which was literally right there 😝

I believe this line in the README confused me

> You'll also need to create/copy secret values for postgres, and any other optional values. See templates/secrets.yaml for possible values

as it lead me to believe both `postgresql-user` and `postgres-password` should be part of the manually created secret for a concourse release.

Now that you removed `postgresql-user` from the concourse secret, the README should state that only `postgresql-uri` is a file we need to create for the secret and only if we don’t enable the postgres subchart.

william-tran commented:
Thanks for proofreading, now that I'm looking through the README there's lots of table entries that need updating. I'll try to get around to this in the next few days.

madmod commented on stable/concourse/README.md:
```diff
@@ -316,19 +301,18 @@ By default, this chart will not use a [Credential Manager](https://concourse.ci/
+    # url:
 
-    ## initial periodic token issued for concourse
-    ## ref: https://www.vaultproject.io/docs/concepts/tokens.html#periodic-tokens
```
I think it might be best to leave this comment as it specifies the kind of token needed. For example if you configured it with a non renewable token it would fail after a period of time and if you use (gasp) a root token it would be a security horror.

madmod commented on stable/concourse/templates/secrets.yaml:
```diff
@@ -9,31 +10,40 @@ metadata:
+  {{- end }}
+  {{- if .Values.credentialManager.vault.enabled }}
+  vault-ca-cert: {{ default "" .Values.secrets.credentialManager.vault.caCert | b64enc | quote }}
+  vault-client-token: {{ default "" .Values.secrets.credentialManager.vault.clientToken | b64enc | quote }}
```
I think it may be incorrect to persist the client token in this way as concourse will need to periodically renew it and redeploying a stale token would cause a failure. (Same for a pod failure or scaling) I haven't verified this so maybe Concourse is smart enough to deal with it and has the latest token stored elsewhere. Vault best practice is to either use an App Role or even better a service account with the Vault Kubernetes auth backend. That has the downside of longer lived credentials, but the advantage of being able to deal with long periods of downtime without requiring manually created tokens. I'm not sure if concourse has a way to save the refreshed tokens but a persistent volume might make more sense.

william-tran replied:
>maybe Concourse is smart enough to deal with it and has the latest token stored elsewhere. 

@madmod I'm not a vault user, but from my reading of https://www.vaultproject.io/docs/concepts/tokens.html#periodic-tokens as long as there isn't downtime longer than the token period, the token value provided would still be valid. So maybe the word `initial` should be dropped from the comment 

```
## initial periodic token issued for concourse
```

Because this value shouldn't change, unless you have enough downtime that the token expires.

madmod commented on stable/concourse/templates/web-deployment.yaml:
```diff
@@ -226,48 +161,24 @@ spec:
-                configMapKeyRef:
-                  name: {{ template "concourse.concourse.fullname" . }}
-                  key: vault-auth-backend
+              value: {{ default "" .Values.credentialManager.vault.authBackend | quote }}
```
Maybe this should be required instead of an empty default to avoid undefined behavior?

william-tran replied:
in the comments for `credentialManager.vault.authBackend`:

```
    ## vault authentication backend, leave this blank if using an initial periodic token
    ## currently supported backends: token, approle, cert.
    ##
    # authBackend:
```

Concourse seems to deal with env vars that are set as empty strings for other features, so unless we can demonstrate otherwise, I'd be hesitant to start wrapping everything in if statements. See also https://concourse.ci/creds.html#vault

madmod commented on stable/concourse/values.yaml:
```diff
@@ -609,3 +456,187 @@ rbac:
+      ## provide the client key for authenticating with the [TLS](https://www.vaultproject.io/docs/auth/cert.html) backend
+      ## the value will be written to /concourse-vault/client.key
+      ## make sure to also set authBackend to `cert`
+      # clientKey: |-
```
Some Vault deployments may have the TCP listener configured to require a client cert and client key alongside any other form of authentication so it is incorrect to make this a choice between the three. It must be client token or app role and/or client cert and client key. [Source](https://www.vaultproject.io/docs/configuration/listener/tcp.html#tls_require_and_verify_client_cert)

It is also possible to use one cert/key pair for the initial connection and another for the auth backend. (Untested but I'm pretty certain based on the documentation.)

An example use case could be having few shared client key pairs granting permission to talk with Vault at all (TCP listener) and a separate client key pair claiming the identity of the client. (TLS auth backend) This would allow for auditing of traffic with TLS decryption but securely prove the identity of the client in a way that could have a chain of trust subject to certificate revocation lists. (A common use case is for short lived ssh certs given to engineers by other engineers. Each engineer is an intermediate CA provided by Vault so provenance of credentials is easy to track.) 

Maybe it would be better to treat this like caCert instead by having it be independent from the auth settings?

madmod reviewed (changes requested):
I think there may be some lifecycle issues with how Vault periodic tokens are being used right now. 

Also using client certs is independent of other auth backends and can actually have separate pairs used for the TCP connection and identity verification. (Which has valid use cases.)

Some of my suggestions may fall outside of the scope of this PR so please let me know if you would like me to create separate issues. See my other comments for details.

unguiculus reviewed (dismissed):
Whenever possible, please add defaults to `values.yaml` instead of using the `default` function.

Commit:
Add missing documentation for values introduced in this and previous PRs

Also fix encryption key values key in secrets, and rename credentialManager.vault.caCert to credentialManager.vault.useCaCert for clarity

william-tran commented:
@unguiculus I'll address the use of `default ""` in a later commit. 

@madmod The documentation for the vault initial periodic token has moved to the bottom of [values.yaml](https://github.com/kubernetes/charts/pull/3254/files#diff-95cc28250f44d26990fee96ad2fbd63dR638) as I've separated out values stored as secrets into their own section.

@viglesiasce How can I become at least a reviewer for this chart? I have a close relationship with the Concourse team so I'd probably be a good candidate to help drive adoption via this chart. Would you sponsor my membership?

viglesiasce commented:
@william-tran I would definitely sponsor your membership!

frodenas committed:
[stable/concourse] bump postgresql dependency to v0.8.11

william-tran commented:
Thanks! I'll have an update this weekend, been slammed lately.

Commit:
Remove redundant default template function calls

william-tran commented:
@viglesiasce @unguiculus  What do you think about rolling all the other outstanding PRs into this one? The only additional functionality is the k8s secrets support #2911 and the postgres version bump #3649

viglesiasce commented:
@william-tran that sounds great to me lets roll it all in here.

william-tran commented:
Sounds like a plan, I will integrate RBAC into #2911 and test on our infra.

william-tran commented:
I was thinking of cutting this PR as-is and leaving k8s secrets support for a followup, but that will require RBAC and given we have an existing value `rbac.serviceAccountName`, I'm thinking we'll need to break that and have `rbac.workerServiceAccountName`and `rbac.webServiceAccountName`. web needs its own service account because it'll use a role binding for get access to secrets, workers don't need that permission. @nexeck since you added RBAC for workers are you ok with this breaking change? I might as well add that breaking change to this PR.

Commit:
Flatten secret values

Use a helper to produce more informative errors when secret values are required

viglesiasce commented:
@william-tran +1 to adding breaking changes here (ie 1.0.0). Where do you stand on whether or not to go with 3.9.0 for this?

Commit:
Add support for using k8s secrets as a credential manager.

Add rbac support for secret access from web pods to to team specific
namespaces.

Upgrade to Concourse 3.9.0

Commit:
Write the liveness probe log to /tmp instead of /concourse-work-dir

If the worker runs out of disk space, logs will stop tee-ing to
/concourse-work-dir/.liveness_probe because it's on the same persistent
volume, and the condition won't be detected.

Commit:
Add OWNERS file

Contributors, reviewers and approvers of PRs to this chart who are also
kubernetes github org members.

william-tran commented:
/assign @viglesiasce

william-tran commented:
I think I'm finally done!

Commit:
fix typos

william-tran commented:
/hold

william-tran commented:
/unhold

Commit:
see if btrfs works in e2e test

Commit:
Stick with naive FS driver since its the most compatible

william-tran commented:
I was of switching the default FS driver to btrfs from naive, as there's some good news that some changes in 3.9 are making btrfs far more stable: https://github.com/concourse/concourse/issues/1966. But looking back at https://github.com/concourse/concourse/issues/1230, I think having the most compatible option as the default is the right thing to do.

william-tran commented:
@unguiculus have I addressed your concerns around the use of `default`?

Commit:
Fix #3407, make docs reflect actual value

Commit:
Remove core maintainers from local OWNERS

The maintainers have OWNER on the top level of the repo already

viglesiasce commented:
/lgtm

Thanks @william-tran!
